Documentation/JSONComparisonMigration.md

```diff
@@ -0,0 +1,330 @@
+event,
+new ValueTypeMatch("timestamp") // Enables value type validation for the specified path
+);
+```
+
+After (option 2: base mode value type validation)
+```java
+String expected = "{ \"timestamp\": \"STRING_TYPE\", \"key1\": 123 }"; // "STRING_TYPE" is just a value convention - the only requirement is that the value type is the one you want to validate
+assertTypeMatch( // Base value comparison mode is type match
+expected,
+event,
+new ValueExactMatch("key1") // Enables exact value validation for the specified path
+);
+```
+
+This pattern is commonly encountered when performing JSON payload validation with randomly generated or time-based values, such as ECIDs or timestamps. By using only value type validation on these values, the correct type can be validated without needing to pre-capture the actual value.
+
+### Collection equal count
+
+#### Key pattern
+
+1. Use of flattened map count.
+2. Use of other path option logic that prevents using [`assertEquals`](#equals-validation).
+
+Before
+```java
+// Example event payload
+Map<String, Object> event = new HashMap<>();
+event.put("timestamp", "2024-08-06T22:45:26Z");
+event.put("key1", 123);
+
+Map<String, Object> flattenedEventData = flattenMap(event);
+assertEquals(2, flattenedEventData.count); // (Key pattern 1.)
+assertEquals(123, flattenedEventData.get("key1"));// (Key pattern 2.) exact value - requiring other path option
+assertNotNull(flattenedEventData.get("timestamp")); // (Key pattern 2.) value type - requiring other path option
+```
+
+After
+```java
+String expected = "{ \"timestamp\": \"STRING_TYPE\", \"key1\": 123 }"; 
+assertExactMatch(
+expected,
+event,
+new CollectionEqualCount(Subtree), // Disables extensible collections from the root of the JSON and all nodes under it
+new ValueTypeMatch("timestamp") // Mix of value exact and type validation prevents using `assertEqual`
+);
+```
+
+This pattern is commonly encountered when explicitly restricting the number of actual elements to match what is expected, as the default validation logic allows for extensible collections.
+
+### Element count validation
+
+#### Key pattern
+
+1. Use of flattened map count without checking each property individually.
+
+Before
+```java
+// Example event payload
+Map<String, Object> event = new HashMap<>();
+event.put("type", "edge");
+event.put("key1", 123);
+
+Map<String, Object> flattenedEventData = flattenMap(event);
+
+assertEquals(2, flattenedEventData.size()); // (Key pattern 1.) flattened map count
+assertEquals("edge", flattenedEventData.get("type")); 
+// (Key pattern 1.) Notice not all properties are covered in the property-by-property assertion
+```
+
+After
+```java
+String expected = "{ \"type\": \"edge\"}"; // Notice only defining expected validation for property that was explicitly covered
+assertExactMatch(
+expected,
+event,
+new ElementCount(2, Subtree) // Element count with 2 total properties in the entire JSON, `Subtree` refers to this node and everything under it
+);
```

Comment:
I changed the comment for this than iOS one, please check.
In iOS version:
// The combination of: 
   // 1. Default `paths` = `nil` -> root of the JSON, AND
   // 2. `scope` = `.subtree` -> this node and everything under it